etcd/datadog_checks/etcd/etcd.py

```diff
@@ -69,7 +70,118 @@ class Etcd(AgentCheck):
     }
 
+    def __init__(self, name, init_config, agentConfig, instances=None):
+        if instances is not None:
+            for instance in instances:
+                # For legacy check ensure prometheus_url is set so
+                # OpenMetricsBaseCheck instantiation succeeds
+                if not is_affirmative(instance.get('use_preview', False)):
+                    instance.setdefault('prometheus_url', '')
+
+        super(Etcd, self).__init__(
+            name,
+            init_config,
+            agentConfig,
+            instances,
+            default_instances={
+                'etcd': {
+                    'prometheus_url': 'http://localhost:2379/metrics',
+                    'namespace': 'etcd',
+                    'metrics': [METRIC_MAP],
+                    'send_histograms_buckets': True,
+                }
+            },
+            default_namespace='etcd',
+        )
+
     def check(self, instance):
+        if is_affirmative(instance.get('use_preview', False)):
+            self.check_post_v3(instance)
+        else:
+            self.warning(
+                'In Agent 6.9 this check will only support ETCD v3+. If you '
+                'wish to preview the new version, set `use_v3` to `true`.'
+            )
+            self.check_pre_v3(instance)
+
+    @classmethod
+    def access_api(cls, scraper_config, path, data='{}'):
+        url = urlparse(scraper_config['prometheus_url'])
+        endpoint = '{}://{}{}'.format(url.scheme, url.netloc, path)
+
+        timeout = scraper_config['prometheus_timeout']
+
+        username = scraper_config['username']
+        password = scraper_config['password']
+        auth = (username, password) if username and password else None
+
+        cert = None
+        if isinstance(scraper_config['ssl_cert'], string_types):
+            if isinstance(scraper_config['ssl_private_key'], string_types):
+                cert = (scraper_config['ssl_cert'], scraper_config['ssl_private_key'])
+            else:
+                cert = scraper_config['ssl_cert']
+
+        verify = True
+        if isinstance(scraper_config['ssl_ca_cert'], string_types):
+            verify = scraper_config['ssl_ca_cert']
+
+        response = {}
+        try:
+            r = requests.post(endpoint, data=data, timeout=timeout, auth=auth, verify=verify, cert=cert)
+            response.update(r.json())
+        except Exception:
+            pass
+
+        return response
+
+    @classmethod
+    def is_leader(cls, scraper_config):
+        # Modify endpoint as etcd stabilizes
+        # https://github.com/etcd-io/etcd/blob/master/Documentation/dev-guide/api_grpc_gateway.md#notes
+        response = cls.access_api(scraper_config, '/v3beta/maintenance/status')
+
+        leader = response.get('leader')
+        member = response.get('header', {}).get('member_id')
+
+        return leader and member and leader == member
+
+    @classmethod
+    def get_leader_state(cls, scraper_config, tags):
```

Comment:
Not a big fan of the name, since the method modifies the tags parameter. I would prefer something more like `add_leader_state_tag`